### ISSUE TITLE
max_instances for a scheduler per job_id instead of per function

### ISSUE THREAD

**Issue by melgo-tb** [agronholm/apscheduler#579]
**Is your feature request related to a problem? Please describe.**
I would want the scheduler not to overlap execution of a particular unique job_id, right now APScheduler does it at the function level and not at the job_id level. For example, if I have created a scheduler that executes a function with a particular job_ids AB123 ,DC321, and max_instances set to  1 the jobs don't execute concurrently even with different unique ids.
My function has a read & update to an external data source and I don't want the next execution to start with stale values but rather start after the previous update is complete.

**Describe the solution you'd like**
I would like an option to set max_instances at a job_id level, where for each particular job_id (& not function) we can define the max_instances

**Describe alternatives you've considered**
I have considered dynamically checking a number of jobs and modifying the max_instances but this has scenarios of overlap of jobs with the same id.

**Additional context**

**Comment by ljluestc** [comment]
import threading
from apscheduler.schedulers.background import BackgroundScheduler
from concurrent.futures import ThreadPoolExecutor
import time

class CustomScheduler:
    def __init__(self):
        self.scheduler = BackgroundScheduler(executor={'default': ThreadPoolExecutor(10)})
        self.job_locks = {}
        self.scheduler.start()

    def add_job(self, func, job_id, *args, **kwargs):
        # Schedule the job with a wrapper that manages concurrency based on job ID
        self.scheduler.add_job(self.job_wrapper, args=(func, job_id) + args, kwargs=kwargs, id=job_id)

    def job_wrapper(self, func, job_id, *args, **kwargs):
        # Create a lock for the job ID if it doesn't exist
        if job_id not in self.job_locks:
            self.job_locks[job_id] = threading.Lock()

        lock = self.job_locks[job_id]

        # Acquire the lock for the specific job ID
        with lock:
            # Run the actual job function
            return func(*args, **kwargs)

# Example usage
def example_job(job_id, data):
    print(f"Running job {job_id} with data: {data}")
    time.sleep(2)  # Simulating a job process

custom_scheduler = CustomScheduler()

# Adding jobs
custom_scheduler.add_job(example_job, "AB123", "AB123", "Data for AB123")
custom_scheduler.add_job(example_job, "DC321", "DC321", "Data for DC321")

# The scheduler will ensure that jobs with the same ID don't run concurrently.

**Comment by agronholm** [comment]
This is already solved in master (v4.0.0a4).

**Comment by dakai-wei-of-shizen** [comment]
I can also confirm that v3.11.0 can run each job_id instead of seeing same function of different id as same job.

here is my example. it is not simplified but as you can understand, A,B,C,D are seen as different jobs, each with a max_instances of 1 and can run at the same time.

```Python
import asyncio
import queue
import sys
import random
import string
from apscheduler.schedulers.asyncio import AsyncIOScheduler

q = queue.Queue()

LIST_SIZE = 100000
produced_data = {"A": [], "B": [], "C": [], "D": []}  # Stores all produced data
consumed_data = {"A": [], "B": [], "C": [], "D": []}  # Stores all consumed data

SCHEDULER_MAX_JOB_LIMIT = 10  # Maximum number of jobs to run concurrently


def generate_random_string(length=10):
    """Generate a random string of fixed length."""
    return "".join(random.choices(string.ascii_letters + string.digits, k=length))


async def producer(name):
    sleep_time = random.uniform(0, 1)  # Sleep for a random duration to cause skip
    await asyncio.sleep(sleep_time)

    """Function for producer threads to add a large list of data to the queue every second."""
    print(f"{name}")
    data = [f"{name}-{i}-{generate_random_string()}" for i in range(LIST_SIZE)]
    produced_data[name].extend(data)  # Store produced data
    q.put((name, data))

    await asyncio.sleep(sleep_time)


def get_queue_size_in_mb(queue):
    """Function to estimate the total memory size of the queue in MB."""
    total_size = 0
    # Iterate through the queue and calculate memory usage of the items
    for item in list(queue.queue):  # We use list(queue.queue) to access the internal list
        name, data = item
        total_size += sys.getsizeof(data)  # Get size of the data list in the item
        for string in data:
            total_size += sys.getsizeof(string)  # Add size of each string in the list
    return total_size / (1024 * 1024)  # Convert to MB


async def consumer():
    """Async function to consume data from the queue and verify all inputs."""

    # Get the current queue size in MB before consuming any data
    queue_size_mb = get_queue_size_in_mb(q)
    print(f"Status: Queue size - {queue_size_mb:.2f} MB")

    # Now consume data from the queue
    while not q.empty():
        name, data = q.get()
        consumed_data[name].extend(data)  # Store consumed data

    # Compare produced and consumed data
    for key in produced_data.keys():
        if len(consumed_data[key]) > 0:  # Ensure there is data to compare
            # Pick a random index in the consumed data
            random_index = random.randint(0, len(consumed_data[key]) - 1)
            consumed_item = consumed_data[key][random_index]
            produced_item = produced_data[key][random_index]

            if consumed_item == produced_item:
                print(f"Data match for {consumed_item}: Consumer and Producer data match.")
            else:
                print(
                    f"Data mismatch for {consumed_item}: Consumer data does not match Producer data!!!!!!!!!!!!!!!!!!!!!!!!!"
                )

    # Print current data sizes
    print(
        f"Status: Produced sizes - {{'A': {len(produced_data['A'])}, 'B': {len(produced_data['B'])}, 'C': {len(produced_data['C'])}, 'D': {len(produced_data['D'])}}}"
    )
    print(
        f"Status: Consumed sizes - {{'A': {len(consumed_data['A'])}, 'B': {len(consumed_data['B'])}, 'C': {len(consumed_data['C'])}, 'D': {len(consumed_data['D'])}}}"
    )

    # Randomly decide to sleep for a longer period (to simulate delays and missed jobs)
    sleep_time = random.uniform(0, 8)  # Sleep for a random duration to cause skip
    await asyncio.sleep(sleep_time)


async def start_producer_jobs(scheduler):
    """Schedule the producer jobs using the scheduler."""
    for name in ["A", "B", "C", "D"]:
        # Assign a unique job ID to each producer job
        job_id = f"producer_{name}"
        scheduler.add_job(producer, "interval", seconds=1, args=(name,), id=job_id)


async def start_consumer_job(scheduler):
    """Schedule the consumer job using the scheduler."""
    scheduler.add_job(consumer, "interval", seconds=5, id="consumer", max_instances=1)


async def main():
    # Create and start the scheduler
    scheduler = AsyncIOScheduler(
        {
            "apscheduler.executors.threadpool": {
                "class": "apscheduler.executors.pool:ThreadPoolExecutor",
            },
        }
    )
    scheduler.start()

    # Schedule producer and consumer jobs
    await start_producer_jobs(scheduler)
    await start_consumer_job(scheduler)

    # Keep the main thread running to allow the scheduler to operate
    while True:
        await asyncio.sleep(1)


if __name__ == "__main__":
    asyncio.run(main())

```
